packages/providers/src/utils/auto-retry-fetch.ts

```diff
@@ -0,0 +1,78 @@
+import type { ProviderOptions } from '../provider';
+
+import { sleep } from './sleep';
+
+type Backoff = 'linear' | 'exponential' | 'fixed';
+
+/**
+ * Retry options scheme
+ */
+export type RetryOptions = {
+  /**
+   * Amount of attempts to retry before failing the call.
+   */
+  maxRetries: number;
+  /**
+   * Backoff strategy to use when retrying. Default is exponential.
+   */
+  backoff?: Backoff;
+  /**
+   * Starting delay for backoff strategy. Default is 150ms.
+   */
+  baseDelay?: number;
+};
+
+/**
+ * Calculate the wait duration for the next retry attempt
+ * @param options - Retry options configuration
+ * @param retryAttemptNum - 1-based, tells which retry attempt is this
+ * @returns Next wait duration
+ */
+export function getWaitDelay(options: RetryOptions, retryAttemptNum: number) {
+  const duration = options.baseDelay ?? 150;
+
+  switch (options.backoff) {
+    case 'linear':
+      return duration * retryAttemptNum;
+    case 'fixed':
+      return duration;
+    case 'exponential':
+    default:
+      return 2 ** (retryAttemptNum - 1) * duration;
```

Comment:
I forgot to mention that `**` is the same as `Math.pow`, while `^` is a [bitwise XOR](https://www.w3schools.com/js/js_bitwise.asp) operator.

Reply:
Yeah, math brain fart from my side.